python/triton/compiler/code_generator.py

```diff
@@ -440,22 +438,24 @@ def visit_FunctionDef(self, node):
         self.builder.set_insertion_point_to_start(entry)
         # visit function body
         self.visit_compound_statement(node.body)
+
         # finalize function
+        assert not self.builder.get_insertion_block().has_terminator()
         if self.ret_type is None or self.ret_type == language.void:
             self.ret_type = language.void
             self.builder.ret([])
         else:
-            # update return type
-            if isinstance(self.ret_type, tuple):
-                self.prototype.ret_types = list(self.ret_type)
-                self.fn.reset_type(self.prototype.to_ir(self.builder))
-            else:
-                self.prototype.ret_types = [self.ret_type]
-                self.fn.reset_type(self.prototype.to_ir(self.builder))
+            self.prototype.ret_types = list(self.ret_type) if isinstance(self.ret_type, tuple) else [self.ret_type]
+            self.fn.reset_type(self.prototype.to_ir(self.builder))
+            self.builder.ret([
+                self.builder.create_undef(ty.to_ir(self.builder))
```

Comment:
This is a little strange, but basically when we have:
```python
if x:
    return a
else:
    return b
```

previously we would end up with the IR:
```mlir
cond_br x, ^bb1, ^bb2

^bb1:
tt.return %a

^bb2:
tt.return %b
```

but now we get the more verbose
```mlir
cond_br x, ^bb1, ^bb2

^bb1:
tt.return %a

^bb2:
tt.return %b

^bb3:
br  ^bb5

^bb4:
br ^bb5

^bb5:
tt.return %undef
```

where bb 3, 4, 5 are dead code and will be cleaned up by the canonicalization pass.